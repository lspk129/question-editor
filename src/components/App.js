import React, { Component } from 'react';
import { Grid, Row, Col, TextInput } from './styles';

class App extends Component {
  state = {
    rows: [
      { id: 'ybvtb', text: '' },
      { id: 'm23n4', text: '' },
    ],
    cols: [
      { id: '12345', text: '' },
      { id: 'awdwa', text: '' },
    ],
  };

  handleAdd = axis => () => {
    const id = Math.random().toString(36).slice(-5);
    this.setState({
      [axis]: [...this.state[axis], { id, text: '' }],
    });
  }

  handleTextChange = axis => (e) => {
    const cells = this.state[axis];
    const cellIndex = cells.findIndex(cell => cell.id === e.target.dataset.input);
    const newState = this.state[axis];
    newState[cellIndex].text = e.target.value;
    this.setState({ [axis]: newState });
  }

  handleDelete = axis => (e) => {
    const cells = this.state[axis];
    const cellIndex = cells.findIndex(cell => cell.id === e.target.dataset[axis]);
    const newCellState = [
      ...cells.slice(0, cellIndex),
      ...cells.slice(cellIndex + 1),
    ];
    this.setState({ [axis]: newCellState });
  }

  renderDelete = rows => rows.map(({ id }) => (
    <Col
      key={`${id}`}
      remove
      data-rows={id}
      onClick={this.handleDelete('rows')}
    >
      del
    </Col>
  ));

  renderRows = rows => rows.map(({ id, text }, index) => (
    <Col text key={id}>
      <TextInput
        size={text === '' ? 6 : text.length + 1}
        data-input={id}
        type={'text'}
        value={text}
        placeholder={`row${index + 1}`}
        onChange={this.handleTextChange('rows')}
      />
    </Col>
  ));

  renderCols = cols => cols.map(({ id, text }, index) => (
    <Row key={id}>
      <Col
        remove
        data-cols={id}
        onClick={this.handleDelete('cols')}
      >
        del
      </Col>
      <Col text key={id}>
        <TextInput
          size={text === '' ? 6 : text.length + 1}
          data-input={id}
          type={'text'}
          value={text}
          placeholder={`col${index + 1}`}
          onChange={this.handleTextChange('cols')}
        />
      </Col>
      {this.state.rows.map(col => <Col key={col.id}><input type="radio" /></Col>)}
    </Row>
  ))

  render() {
    return (
      <Grid size={this.state.cols.length + 4}>
        <Row>
          <Col space />
          <Col space />
          {this.renderDelete(this.state.rows)}
          <Col add onClick={this.handleAdd('rows')} >add</Col>
        </Row>
        <Row>
          <Col space />
          <Col space />
          {this.renderRows(this.state.rows)}
        </Row>
        {this.renderCols(this.state.cols)}
        <Row>
          <Col add onClick={this.handleAdd('cols')} >add</Col>
        </Row>
      </Grid>
    );
  }
}

export default App;
