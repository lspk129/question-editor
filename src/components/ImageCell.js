import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const ImageLabel = styled.label`
  border: 1px solid #ccc;
  width: 100%;
  height: 100%;
  cursor: pointer;
  display: grid;
  align-items: center;
  justify-items: center;
  margin: auto;
  font-size: 12px;
`;

const ImageInput = styled.input`
  display: none;
`;

export const Image = styled.img`
  width: 50px;
  height: 50px;
  align-self: center;
  justify-self: center;
`;

export const ImageCell = ({ change, id }) => (
  <ImageLabel>
    <ImageInput
      type={'file'}
      onChange={change}
      data-img={id}
    />
      img
  </ImageLabel>
);

ImageCell.propTypes = {
  change: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
};
