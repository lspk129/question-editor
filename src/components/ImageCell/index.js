import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const ImageLabel = styled.label`
  width: 100%;
  height: 100%;
  cursor: pointer;
  display: grid;
  align-items: center;
  justify-items: center;
  margin: auto;
  font-size: 12px;
  ${({ icon }) => icon && `
    background: url(${icon});
    background-size: 30px 30px;
    background-repeat: no-repeat;
    background-position: center;
    cursor: pointer;
    opacity: 0.7;
  `};
`;

const ImageInput = styled.input`
  display: none;
`;

export const Image = styled.img`
  width: 50px;
  height: 50px;
  align-self: center;
  justify-self: center;
  ${({ small }) => small && `
    width: 30px;
    padding: 0 10px 0 20px;
    opacity: 0.7;
  `}
`;

export const ImageCell = ({ change, id, icon }) => (
  <ImageLabel icon={icon} >
    <ImageInput
      type={'file'}
      onChange={change}
      data-img={id}
    />
  </ImageLabel>
);

ImageCell.propTypes = {
  change: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
  icon: PropTypes.string.isRequired,
};
