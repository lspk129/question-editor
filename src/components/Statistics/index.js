import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import getStatistics from '../../utils';

const Section = styled.div`
  padding: 0 20px 20px;
  ${({ stats }) => stats && `
    padding-top: 50px;
  `};
`;

const Statistics = ({ state }) => {
  const {
    rowNumber,
    colNumber,
    imgNumber,
    longestRowLabel,
    longestColLabel,
  } = getStatistics(state);

  return (
    <Section>
      <h2>Question Summary View</h2>
      <h3>Summary</h3>
      <Section stats>
        <p>Number of rows: {rowNumber}</p>
        <p>Number of columns: {colNumber}</p>
        <p>Number of images uploaded: {imgNumber}</p>
        <p>Longest row label: {longestRowLabel}</p>
        <p>Longest column label: {longestColLabel}</p>
      </Section>
    </Section>
  );
};

Statistics.propTypes = {
  state: PropTypes.shape({
    rows: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.string.isRequired,
      text: PropTypes.string.isRequired,
    })).isRequired,
  }).isRequired,
};

export default Statistics;
