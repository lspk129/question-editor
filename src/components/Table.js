import styled from 'styled-components';

export const TableGrid = styled.div`
  display: grid;
  ${({ size }) => `grid-template-columns: repeat(${size}, max-content)`};
  grid-gap: 10px;
  border-right: 1px solid grey;
`;

export const Row = styled.div`
  display: grid;
  grid-auto-rows: 50px;
  grid-auto-columns: minmax(50px, max-content);
  grid-column-gap: 10px;
  grid-gap: 10px;
`;

export const Col = styled.div`
  display: grid;
  align-items: center;
  justify-items: center;
  font-size: 12px;
  ${({ remove }) => remove && 'background-color: salmon; cursor: pointer'};
  ${({ add }) => add && 'background-color: lightSeaGreen; cursor: pointer'};
  ${({ text }) => text && 'background-color: LightSteelBlue'};
  ${({ space }) => space && 'background-color: #fff'};
  ${({ icon }) => icon && `
    background: url(${icon});
    background-size: 30px 30px;
    background-repeat: no-repeat;
    background-position: center;
    cursor: pointer;
    opacity: 0.7;
  `};
`;

export const TextInput = styled.input`
  text-align: center;
  background: transparent;
  box-sizing: border-box;
  outline: none;
  border: none;
  &::placeholder {
    color: #000;
    font-style: italic;
  }
  ${({ question }) => question && `
    font-size: 20px;
    width: 100%;
    text-align: left;
    &::placeholder {
      color: grey;
    }
  `}
`;

export const Icon = styled.div`
  align-self: center;
  justify-self: center;
`;

export const TableSection = styled.div`
  display: grid;
  ${({ title }) => title && 'padding: 0 20px 20px;'};
`;
