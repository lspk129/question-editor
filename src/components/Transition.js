import styled from 'styled-components';
import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
import './style.css';

/**
 * Normaly I would write props into one styled component
 * but adding props to external component like ReactCSSTransitionGroup
 * getting errors of 'unknown props to <span>' or other tag.
 * Also adding css styles from external sheet, because
 * transitions through styled components do not want to work properly, probobly because
 * styled components styles overides transitions styles
 */
const Transition = styled(ReactCSSTransitionGroup).attrs({
  transitionName: 'example',
  transitionEnterTimeout: 500,
  transitionLeaveTimeout: 300,
})``;

export const TransitionAll = styled(Transition)`
  all: inherit;
`;

export const TransitionRow = styled(TransitionAll)`
  grid-row: 4;
`;

export const TransitionCol = styled(Transition)`
  grid-auto-flow: column;
  display: grid;
  grid-gap: 10px;
  justify-items: center;
`;
