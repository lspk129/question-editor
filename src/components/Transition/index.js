import styled from 'styled-components';
import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
import '../style.css';

/**
 * Adding props to external component like ReactCSSTransitionGroup
 * throws not critical errors of 'unknown prop to...' html tag,
 * a workaround is to extend styles.
 * Transition styles do not respond written in styled-components.
 */
const Transition = styled(ReactCSSTransitionGroup).attrs({
  transitionName: 'example',
  transitionEnterTimeout: 500,
  transitionLeaveTimeout: 300,
})``;

export const TransitionAll = styled(Transition)`
  all: inherit;
`;

export const TransitionRow = styled(TransitionAll)`
  grid-row: 4;
`;

export const TransitionCol = styled(Transition)`
  grid-auto-flow: column;
  display: grid;
  grid-gap: 10px;
  justify-items: center;
`;
