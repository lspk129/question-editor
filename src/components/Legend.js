import React from 'react';
import styled from 'styled-components';
import { Image } from './ImageCell';
import deleteIcon from '../icons/ic_delete_forever_black_24px.svg';
import imageIcon from '../icons/ic_image_black_24px.svg';
import addIcon from '../icons/ic_add_circle_black_24px.svg';

const Section = styled.div`
  grid-column: 2;
  padding: 0 20px 20px;
`;

const ImageRow = styled.p`
  display: flex;
  align-items: center;
  margin: 0;
  ${({ note }) => note && `
    padding: 20px 0 0 30px;
  `};
`;

const Text = styled.span`
  padding-left: 10px;
`;

const Legend = () => (
  <Section>
    <h3>Legend</h3>
    <ImageRow><Image small src={imageIcon} /> - <Text>Add images</Text></ImageRow>
    <ImageRow><Image small src={addIcon} /> - <Text>Add row/column</Text></ImageRow>
    <ImageRow><Image small src={deleteIcon} /> - <Text>Delete row/column</Text></ImageRow>
    <ImageRow note><i>Italic</i> <Text>text is editable</Text></ImageRow>
  </Section>
);

export default Legend;
