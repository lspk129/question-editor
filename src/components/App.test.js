import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import App from './App';
import Statistics from './Statistics';
import Legend from './Legend';
import { ImageCell } from './ImageCell';
import { TableGrid, Row, Col, TextInput, TableSection } from './Table';
import { TransitionAll, TransitionRow, TransitionCol } from './Transition';

describe('App component', () => {
  const wrapper = shallow(<App />);

  it('renders corretly', () => {
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  });

  it('renders Statistics component', () => {
    expect(wrapper.find(Statistics)).toBePresent();
  });

  it('renders Legend component', () => {
    expect(wrapper.find(Legend)).toBePresent();
  });

  it('renders ImageCell component', () => {
    expect(wrapper.find(ImageCell)).toBePresent();
    expect(wrapper.find(ImageCell).length).toEqual(4);
  });

  it('renders TableGrid component', () => {
    expect(wrapper.find(TableGrid)).toBePresent();
    expect(wrapper.find(TableGrid).length).toEqual(1);
  });

  it('renders Row component', () => {
    expect(wrapper.find(Row)).toBePresent();
    expect(wrapper.find(Row).length).toEqual(6);
  });

  it('renders Col component', () => {
    expect(wrapper.find(Col)).toBePresent();
    expect(wrapper.find(Col).length).toEqual(14);
  });

  it('renders TextInput component', () => {
    expect(wrapper.find(TextInput)).toBePresent();
    expect(wrapper.find(TextInput).length).toEqual(5);
  });

  it('renders TableSection component', () => {
    expect(wrapper.find(TableSection)).toBePresent();
    expect(wrapper.find(TableSection).length).toEqual(2);
  });

  it('renders TransitionAll component', () => {
    expect(wrapper.find(TransitionAll)).toBePresent();
    expect(wrapper.find(TransitionAll).length).toEqual(2);
  });

  it('renders TransitionRow component', () => {
    expect(wrapper.find(TransitionRow)).toBePresent();
    expect(wrapper.find(TransitionRow).length).toEqual(3);
  });

  it('renders TransitionCol component', () => {
    expect(wrapper.find(TransitionCol)).toBePresent();
    expect(wrapper.find(TransitionCol).length).toEqual(1);
  });
});
