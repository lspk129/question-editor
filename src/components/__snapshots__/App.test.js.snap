// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders corretly 1`] = `
<styled.div>
  <styled.div
    main={true}
  >
    <styled.div
      title={true}
    >
      <h2>
        Question Editor
      </h2>
      <styled.input
        onChange={[Function]}
        placeholder="Enter title of the question"
        question={true}
        size={30}
        type="text"
        value=""
      />
    </styled.div>
    <styled.div
      size={6}
    >
      <styled.div>
        <Styled(Styled(Styled(CSSTransitionGroup)))>
          <styled.div
            data-rows="ybvtb"
            icon="ic_delete_forever_black_24px.svg"
            onClick={[Function]}
          />
          <styled.div
            data-rows="m23n4"
            icon="ic_delete_forever_black_24px.svg"
            onClick={[Function]}
          />
          <styled.div
            add={true}
            icon="ic_add_circle_black_24px.svg"
            onClick={[Function]}
          />
        </Styled(Styled(Styled(CSSTransitionGroup)))>
      </styled.div>
      <styled.div>
        <Styled(Styled(Styled(CSSTransitionGroup)))>
          <Component
            change={[Function]}
            icon="ic_image_black_24px.svg"
            id="ybvtb"
          />
          <Component
            change={[Function]}
            icon="ic_image_black_24px.svg"
            id="m23n4"
          />
        </Styled(Styled(Styled(CSSTransitionGroup)))>
      </styled.div>
      <styled.div>
        <Styled(Styled(Styled(CSSTransitionGroup)))>
          <styled.div
            text={true}
          >
            <styled.input
              data-input="ybvtb"
              onChange={[Function]}
              placeholder="row1"
              size={6}
              type="text"
              value=""
            />
          </styled.div>
          <styled.div
            text={true}
          >
            <styled.input
              data-input="m23n4"
              onChange={[Function]}
              placeholder="row2"
              size={6}
              type="text"
              value=""
            />
          </styled.div>
        </Styled(Styled(Styled(CSSTransitionGroup)))>
      </styled.div>
      <Styled(Styled(CSSTransitionGroup))>
        <styled.div>
          <styled.div
            data-cols="12345"
            icon="ic_delete_forever_black_24px.svg"
            onClick={[Function]}
          />
          <Component
            change={[Function]}
            icon="ic_image_black_24px.svg"
            id="12345"
          />
          <styled.div
            text={true}
          >
            <styled.input
              data-input="12345"
              onChange={[Function]}
              placeholder="col1"
              size={6}
              type="text"
              value=""
            />
          </styled.div>
          <Styled(Styled(CSSTransitionGroup))>
            <styled.div>
              <input
                checked={false}
                data-radio="ybvtb"
                onClick={[Function]}
                type="radio"
              />
            </styled.div>
            <styled.div>
              <input
                checked={false}
                data-radio="m23n4"
                onClick={[Function]}
                type="radio"
              />
            </styled.div>
          </Styled(Styled(CSSTransitionGroup))>
        </styled.div>
        <styled.div>
          <styled.div
            data-cols="awdwa"
            icon="ic_delete_forever_black_24px.svg"
            onClick={[Function]}
          />
          <Component
            change={[Function]}
            icon="ic_image_black_24px.svg"
            id="awdwa"
          />
          <styled.div
            text={true}
          >
            <styled.input
              data-input="awdwa"
              onChange={[Function]}
              placeholder="col2"
              size={6}
              type="text"
              value=""
            />
          </styled.div>
          <Styled(Styled(CSSTransitionGroup))>
            <styled.div>
              <input
                checked={false}
                data-radio="ybvtb"
                onClick={[Function]}
                type="radio"
              />
            </styled.div>
            <styled.div>
              <input
                checked={false}
                data-radio="m23n4"
                onClick={[Function]}
                type="radio"
              />
            </styled.div>
          </Styled(Styled(CSSTransitionGroup))>
        </styled.div>
      </Styled(Styled(CSSTransitionGroup))>
      <styled.div>
        <styled.div
          add={true}
          icon="ic_add_circle_black_24px.svg"
          onClick={[Function]}
        />
      </styled.div>
    </styled.div>
  </styled.div>
  <Statistics
    state={
      Object {
        "cols": Array [
          Object {
            "id": "12345",
            "img": Array [],
            "text": "",
          },
          Object {
            "id": "awdwa",
            "img": Array [],
            "text": "",
          },
        ],
        "rows": Array [
          Object {
            "checked": Array [
              false,
              false,
            ],
            "id": "ybvtb",
            "img": Array [],
            "text": "",
          },
          Object {
            "checked": Array [
              false,
              false,
            ],
            "id": "m23n4",
            "img": Array [],
            "text": "",
          },
        ],
        "title": "",
      }
    }
  />
  <Legend />
</styled.div>
`;
